File: src/wasm-projectors/rust/sekiban-mv/src/projector.rs
//! Trait a WASM-side materialized view projector implements.
//!
//! Mirrors the C# `IWasmMvProjector` interface one-for-one so the `ClassRoomEnrollmentMvV1`
//! implementations stay structurally identical between languages.

use crate::dto::{MvSerializableEventDto, MvSqlStatementDto, MvTableBindingsDto};
use crate::query_port::MvQueryPort;

pub trait WasmMvProjector: Send + Sync + 'static {
    fn view_name(&self) -> &'static str;
    fn view_version(&self) -> i32;
    fn logical_tables(&self) -> &'static [&'static str];

    fn initialize(&self, tables: &MvTableBindingsDto) -> Vec<MvSqlStatementDto>;

    fn apply_event(
        &self,
        tables: &MvTableBindingsDto,
        event: &MvSerializableEventDto,
        query_port: &dyn MvQueryPort,
    ) -> Vec<MvSqlStatementDto>;
}

File: src/lib/sekiban-go/mv/projector.go
package mv

// QueryPort is the minimal interface a projector uses to consult already-projected rows
// mid-apply. The default HostBackedQueryPort routes through the env.mv_host_query_rows host
// import; projectors that do not query can ignore the argument entirely.
type QueryPort interface {
	QueryRows(sql string, params []MvParam) []MvQueryRowDto
	QuerySingleRow(sql string, params []MvParam) *MvQueryRowDto
}

// Projector is what each materialized-view implementation satisfies. Mirrors the Rust
// `WasmMvProjector` trait and the Swift protocol.
type Projector interface {
	ViewName() string
	ViewVersion() int32
	LogicalTables() []string

	Initialize(tables MvTableBindingsDto) []MvSqlStatementDto
	ApplyEvent(
		tables MvTableBindingsDto,
		event MvSerializableEventDto,
		queryPort QueryPort,
	) []MvSqlStatementDto
}
